src/wasm-type.h

```diff
@@ -396,6 +397,37 @@ struct Rtt {
   std::string toString() const;
 };
 
+// TypeBuilder - allows for the construction of recursive types. Contains a
+// table of `n` mutable HeapTypes and can construct temporary types that are
+// backed by those HeapTypes, refering to them by reference. Those temporary
+// types are owned by the TypeBuilder and should only be used in the
+// construction of HeapTypes to insert into the TypeBuilder. Temporary types
+// should never be used in the construction of normal Types, only other
+// temporary types.
+struct TypeBuilder {
+  std::vector<HeapType> heapTypes;
+
+  struct Impl;
+  std::unique_ptr<Impl> impl;
+
+  TypeBuilder(size_t n);
+  ~TypeBuilder();
+
+  TypeBuilder(TypeBuilder& other) = delete;
+  TypeBuilder(TypeBuilder&& other) = delete;
+  TypeBuilder& operator=(TypeBuilder&) = delete;
+
+  HeapType& operator[](size_t i) { return heapTypes[i]; }
+
+  Type getTempTupleType(const Tuple&);
+  Type getTempRefType(size_t i, bool nullable);
```

Comment:
A comment explaining what `i` is might help. Reading this the first time, I have no guesses as to what it could be.